pkg/ring/ring.go

```diff
@@ -461,6 +462,101 @@ func (r *Ring) Subring(key uint32, n int) ReadRing {
 	return sub
 }
 
+// ShuffleShard returns a subring for the provided identifier (eg. a tenant ID)
+// and size (number of instances). The size is expected to be a multiple of the
+// number of zones and the returned subring will contain the same number of
+// instances per zone as far as there are enough registered instances in the ring.
+//
+// The algorithm used to build the subring is a shuffle sharder based on probabilistic
+// hashing. We treat each zone as a separate ring and pick N unique replicas from each
+// zone, walking the ring starting from random but predictable numbers. The random
+// generator is initialised with a seed based on the provided identifier.
+//
+// This implementation guarantees:
+// - Stability: given the same ring, two invocations returns the same result.
+// - Consistency: adding/removing 1 instance from the ring generates a resulting
+//   subring with no more then 1 difference.
+// - Shuffling: probabilistically, for a large enough cluster each identifier gets
+//   a different set of instances, with a reduced number of overlapping instances
+//   between two identifiers.
+func (r *Ring) ShuffleShard(identifier string, size int) ReadRing {
+	// Nothing to do if the shard size is not smaller then the actual ring.
+	if size <= 0 || r.IngesterCount() <= size {
+		return r
+	}
+
+	// Use the identifier to compute an hash we'll use to seed the random.
+	hasher := md5.New()
+	hasher.Write([]byte(identifier)) // nolint:errcheck
+	checksum := hasher.Sum(nil)
+
+	// Generate the seed based on the first 64 bits of the checksum.
+	seed := int64(binary.BigEndian.Uint64(checksum))
+
+	// Initialise the random generator used to select instances in the ring.
+	random := rand.New(rand.NewSource(seed))
+
+	r.mtx.RLock()
+
+	// We expect the shard size to be divisible by the number of zones, in order to
+	// have nodes balanced across zones. If it's not, we do round up.
+	numInstancesPerZone := int(math.Ceil(float64(size) / float64(len(r.ringZones))))
```

Comment:
Just want to check: in the case where we're not using zones, zone = ingesterID.  Will this algorithm end distributing all users to all ingesters in this case?

Reply:
> in the case where we're not using zones, zone = ingesterID

It was in the initial implementation but was soon fixed in #2404. When not using zones, the `zone` is an empty string, which works fine here (all ingesters belong to the same zone).

However, I'm wondering if we should  add a boolean flag to the ring to explicitly enable zone awareness, disabled by default instead of just relying on the fact zones are set (think about a rolling update adding zones to ingesters, you may want to rollout all ingesters first and then enable it).

Reply:
Agree @ranton256. Will do it in a separate PR (since it's unrelated from the work done in this PR).